src/merge_data.rs

```diff
@@ -0,0 +1,293 @@
+}
+
+impl MergeDeferredDataSource {
+    pub fn new(data_sources: Vec<Box<dyn DeferredDataSource>>) -> Self {
+        assert!(!data_sources.is_empty());
+        let mut infos = Vec::new();
+        infos.resize_with(data_sources.len(), VecDeque::new);
+        Self {
+            data_sources,
+            infos,
+            mapping: Vec::new(),
+        }
+    }
+
+    fn merge_entry(first: EntryInfo, second: EntryInfo) -> EntryInfo {
+        let EntryInfo::Panel {
+            short_name,
+            long_name,
+            summary: first_summary,
+            mut slots,
+        } = first
+        else {
+            unreachable!();
+        };
+
+        let EntryInfo::Panel {
+            summary: second_summary,
+            slots: second_slots,
+            ..
+        } = second
+        else {
+            unreachable!();
+        };
+
+        assert!(first_summary.is_none());
+        assert!(second_summary.is_none());
+
+        slots.extend(second_slots);
+
+        EntryInfo::Panel {
+            short_name,
+            long_name,
+            summary: None,
+            slots,
+        }
+    }
+
+    fn compute_mapping(source_infos: &[DataSourceInfo]) -> Vec<u64> {
+        // Compute the mapping from old to new entries (basically the offset
+        // of each initial slot)
+        let slot_lens: Vec<_> = source_infos
+            .iter()
+            .map(|info| {
+                let EntryInfo::Panel { ref slots, .. } = info.entry_info else {
+                    unreachable!();
+                };
+                slots.len() as u64
+            })
+            .collect();
+
+        let mut mapping = Vec::new();
+        let mut sum = 0;
+        for slot_len in slot_lens {
+            mapping.push(sum);
+            sum += slot_len;
+        }
+
+        mapping
+    }
+
+    fn merge_infos(source_infos: Vec<DataSourceInfo>) -> DataSourceInfo {
+        assert!(!source_infos.is_empty());
+
+        // Some fields can't be meaningfully merged, so just assert they're
+        // all equivalent.
+        let first_info = source_infos.first().unwrap();
+        let tile_set = first_info.tile_set.clone();
+        let field_schema = first_info.field_schema.clone();
+
+        for info in &source_infos {
+            assert_eq!(tile_set, info.tile_set);
+            assert_eq!(field_schema, info.field_schema);
+        }
+
+        // Merge remaining fields
+        // IMPORTANT: entry_info must be kept consistent with compute_mapping
+        let interval = source_infos
+            .iter()
+            .map(|info| info.interval)
+            .reduce(Interval::union)
+            .unwrap();
+        let entry_info = source_infos
+            .iter()
+            .map(|info| info.entry_info.clone())
+            .reduce(Self::merge_entry)
+            .unwrap();
+
+        DataSourceInfo {
+            entry_info,
+            interval,
+            tile_set,
+            field_schema,
+        }
+    }
```

Comment:
Thoughts on basic unit tests for functions like this? They seem amenable to unit testing but maybe relying  on the compiler for assurances is more common in Rust-land?

Reply:
Look at the most recent push and see if it seems reasonable.

In general, Rust users do rely a lot on the type system to make sure things work. And Rust's type system is pretty good, so I have a lot of confidence when things compile. But there are still classes of logic errors we hit, so it would be better to add more unit testing in general. So thanks for the reminder.